### PR TITLE
feat(ui): support advanced proxy cache configurations in UI

### PR DESCRIPTION
**Pull request by stonezdj**
- Support repository filter configuration (pattern and kind) for proxy cache projects.
- Ensure the local on not found and proxy referrer API checkbox editability is controlled by project configuration update permissions.

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #(issue)

Please indicate you've done the following:
- [ ] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [ ] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

**Comment by wy65701436** [comment]
cc @bupd and @Vad1mo could you please help to review?

**Review by bupd — CHANGES_REQUESTED** [review]
Image

please fix CI

**Comment by Vad1mo** [comment]
Comparing this against the merged proposal (goharbor/community#280) and the backend implementation (#23527), a few deviations:

**1. Client-side validation is missing**

The proposal's UI section requires: *"Validation should happen both client-side and server-side."*

Neither the filter input in `create-project.component.html` nor the one in `project-policy-config.component.html` has any validator — an invalid pattern is only caught by the server's 400 on save. At minimum the regex kind can be validated client-side (`try { new RegExp(pattern) }`), and a doublestar pattern can be checked for balanced brackets/braces.

**2. Tooltip describes anchoring semantics the backend doesn't have**

`REPOSITORY_FILTER_TOOLTIP` (all 10 locales) says *"regular expression patterns are head-anchored"*. The backend (#23527) requires the pattern to match the **entire** repository path, and the proposal says regex is *"always auto-anchored"*. A user following this tooltip writes `^library/` and matches nothing. The tooltip should say the pattern must match the full repository path, e.g. `^library/.*`.

**3. `allowUpdateProxyCacheConfiguration` binding is dead**

The proxy cache checkbox changes from hardcoded `disabled` to `[disabled]="!allowUpdateProxyCacheConfiguration"`, but this PR removes the only assignment to that field. It's initialized to `false` (`project-policy-config.component.ts:116`) and never set, so the checkbox is still permanently disabled — the dynamic-looking binding just obscures that. Either restore an assignment or keep the hardcoded `disabled`.

Related: sibling controls in the same section now gate on three different signals — the filter input/select on `hasChangeConfigRole`, the enable checkbox on the dead `allowUpdateProxyCacheConfiguration`, and section visibility on `isSystemAdmin` (previously `hasChangeConfigRole`). The `isSystemAdmin` switch also changes who can see the proxy cache section, which the proposal doesn't cover.

**4. Dead component state**

`repositoryFilterPattern`/`repositoryFilterKind` are declared and populated in `retrieve()`, but the template binds `projectPolicy.ProxyCacheFilterPattern`/`ProxyCacheFilterKind` — the two fields are never read and can be removed.

**5. Scope beyond the proposal**

The proposal's UI section only asks for the Repository filter (label, kind selector, pattern input) in the create view. This PR additionally reworks the referrer API checkbox layout and permission gating, switches the proxy-cache section gating to `isSystemAdmin`, and adds a read-only filter display for non-admin members. These may be worthwhile, but they're not spec-driven — worth calling out in the PR description (the current second bullet covers only part of this) or splitting out.

**Minor:** `updateProjectPolicy` in `project.service.ts` now sends `proxy_cache_filter_pattern`/`proxy_cache_filter_kind` on every project update, including non-proxy projects (empty string + `doublestar` in the else branch). The backend ignores them for non-proxy projects on this endpoint, but sending them unconditionally relies on that. Also: `REPOSITORY_FILTER_KIND_*` constants are defined in `base/project/project.ts` and used in create-project, while this component and the service hardcode `'doublestar'` — one or the other.

**Comment by Vad1mo** [comment]
behavior changes for existing users:

1. Project admins/maintainers lose visibility of the proxy cache config. On main, the proxy-cache checkbox block and the endpoint/bandwidth/connection-limit block are gated on hasChangeConfigRole (template lines 19, 36), so project-level admins see them (read-only). The PR regates both to isSystemAdmin. A non-sysadmin project admin now sees nothing about proxy cache except the new read-only filter row, and that only when a filter is actually set.
2. Referrer API becomes editable after project creation. On main the checkbox is [disabled]="!allowUpdateProxyCacheConfiguration", and that flag is provably always false in this view: it's hasChangeConfigRole && !isProxyCacheProject, and isProxyCacheProject is true whenever the project has a registry_id (project-config.component.ts:50-52) — which is exactly when the section renders. So no one could change it post-creation. The PR rebinds it to !hasChangeConfigRole, making it mutable by sysadmins. This is the PR body's stated intent, but it flips a previously-immutable setting to mutable. Side note: the same PR-body claim about local-on-not-found is a no-op — that checkbox is already gated by !hasChangeConfigRole on main (line 164).

**Comment by wy65701436** [comment]
can you give a screenshot for this change?

**Comment by bupd** [comment]
@wy65701436 @stonezdj 

I dont know what makes this review a low quality one 

Screenshot_2026-07-24_19-28-16

**Comment by bupd** [comment]
The UI Unit Tests are clearly failing and its the authors duty to fix and make sure ci is passing

**Review by bupd — DISMISSED** [review]
@stonezdj Thanks for your contribution

suggestion: Please fix the lint issues given below


```
08:05:18 + npm run lint
08:05:18
08:05:18 > harbor@2.15.0 lint
08:05:18 > ng lint
08:05:18
08:05:19
08:05:19 Linting "harbor-portal"...
08:05:32
08:05:32 /home/ubuntu/_work/harbor/harbor/src/github.com/goharbor/harbor/src/portal/src/app/base/left-side-nav/projects/create-project/create-project.component.html
08:05:32 214:35 error Replace `·class="clr-control-label·endpoint·required">{{` with `⏎································class="clr-control-label·endpoint·required"` prettier/prettier
08:05:32 215:33 error Replace `'PROJECT.ENDPOINT'·|·translate` with `>{{·'PROJECT.ENDPOINT'·|·translate·}}</label` prettier/prettier
08:05:32 216:29 error Delete `}}</label` prettier/prettier
08:05:32
08:05:32 /home/ubuntu/_work/harbor/harbor/src/github.com/goharbor/harbor/src/portal/src/app/base/left-side-nav/projects/create-project/create-project.component.ts
08:05:32 401:54 error Delete `⏎················` prettier/prettier
08:05:32
08:05:32 ✖ 4 problems (4 errors, 0 warnings)
08:05:32 4 errors and 0 warnings potentially fixable with the `--fix` option.
08:05:32
08:05:32 Lint errors found in the listed files.
08:05:32 
Error: Process completed with exit code 1.
```

fixing lint might probably fix failing ci workflow  https://github.com/goharbor/harbor/actions/runs/29078451168/job/86315411475

**Comment by wy65701436** [comment]
> @wy65701436 @stonezdj
> 
> I dont know what makes this review a low quality one
> Screenshot_2026-07-24_19-28-16

I'm not sure who added the comments, cc @Vad1mo and @stonezdj do you have any idea?
And I agree that the CI must be passed.

**Comment by stonezdj** [comment]
> > @wy65701436 @stonezdj
> > I dont know what makes this review a low quality one
> > Screenshot_2026-07-24_19-28-16
> 
> I'm not sure who added the comments, cc @Vad1mo and @stonezdj do you have any idea? And I agree that the CI must be passed.

This PR is currently a work in progress, so errors are bound to happen. Outside of open source, I have other full-time commitments. This is an open and inclusive community, so please keep comments respectful and constructive! @wy65701436 @bupd @Vad1mo

**Comment by stonezdj** [comment]
The screenshot of the UI
1.User can specify the filter when creating proxy cache project.
Screenshot 2026-07-30 at 16 45 24
2.The proxy cache filter can be updated for a proxy cache project.
Screenshot 2026-07-30 at 16 46 17

**Comment by stonezdj** [comment]
> Comparing this against the merged proposal ([goharbor/community#280](https://github.com/goharbor/community/pull/280)) and the backend implementation (#23527), a few deviations:
> 
> **1. Client-side validation is missing**
> 
> The proposal's UI section requires: _"Validation should happen both client-side and server-side."_
> 
> Neither the filter input in `create-project.component.html` nor the one in `project-policy-config.component.html` has any validator — an invalid pattern is only caught by the server's 400 on save. At minimum the regex kind can be validated client-side (`try { new RegExp(pattern) }`), and a doublestar pattern can be checked for balanced brackets/braces.
> 
> **2. Tooltip describes anchoring semantics the backend doesn't have**
> 
> `REPOSITORY_FILTER_TOOLTIP` (all 10 locales) says _"regular expression patterns are head-anchored"_. The backend (#23527) requires the pattern to match the **entire** repository path, and the proposal says regex is _"always auto-anchored"_. A user following this tooltip writes `^library/` and matches nothing. The tooltip should say the pattern must match the full repository path, e.g. `^library/.*`.
> 
> **3. `allowUpdateProxyCacheConfiguration` binding is dead**
> 
> The proxy cache checkbox changes from hardcoded `disabled` to `[disabled]="!allowUpdateProxyCacheConfiguration"`, but this PR removes the only assignment to that field. It's initialized to `false` (`project-policy-config.component.ts:116`) and never set, so the checkbox is still permanently disabled — the dynamic-looking binding just obscures that. Either restore an assignment or keep the hardcoded `disabled`.
> 
> Related: sibling controls in the same section now gate on three different signals — the filter input/select on `hasChangeConfigRole`, the enable checkbox on the dead `allowUpdateProxyCacheConfiguration`, and section visibility on `isSystemAdmin` (previously `hasChangeConfigRole`). The `isSystemAdmin` switch also changes who can see the proxy cache section, which the proposal doesn't cover.
> 
> **4. Dead component state**
> 
> `repositoryFilterPattern`/`repositoryFilterKind` are declared and populated in `retrieve()`, but the template binds `projectPolicy.ProxyCacheFilterPattern`/`ProxyCacheFilterKind` — the two fields are never read and can be removed.
> 
> **5. Scope beyond the proposal**
> 
> The proposal's UI section only asks for the Repository filter (label, kind selector, pattern input) in the create view. This PR additionally reworks the referrer API checkbox layout and permission gating, switches the proxy-cache section gating to `isSystemAdmin`, and adds a read-only filter display for non-admin members. These may be worthwhile, but they're not spec-driven — worth calling out in the PR description (the current second bullet covers only part of this) or splitting out.
> 
> **Minor:** `updateProjectPolicy` in `project.service.ts` now sends `proxy_cache_filter_pattern`/`proxy_cache_filter_kind` on every project update, including non-proxy projects (empty string + `doublestar` in the else branch). The backend ignores them for non-proxy projects on this endpoint, but sending them unconditionally relies on that. Also: `REPOSITORY_FILTER_KIND_*` constants are defined in `base/project/project.ts` and used in create-project, while this component and the service hardcode `'doublestar'` — one or the other.

@Vad1mo I have fixed the 1-4 issue, for 5. I enabled the configure change related to update the proxy filter and the serve stale content when upstream unreachable. remove the restriction on the UI should make sense to the user, or they have to delete the proxy cache project and re-create it. update configuration is more straight forward.

**Comment by wy65701436** [comment]
> constructive

Please mark any issue you're working on as **WIP**, so other maintainers know to hold off on reviewing your changes until they're ready.

File: src/portal/src/i18n/lang/ru-ru-lang.json [hunk 1]
```diff
@@ -421,7 +421,14 @@
         "PROXY_CACHE_LOCAL_ON_NOT_FOUND": "Отдавать кэшированный контент при недоступности вышестоящего реестра",
         "PROXY_CACHE_LOCAL_ON_NOT_FOUND_TOOLTIP": "Включите эту опцию, чтобы разрешить проекту продолжать отдавать кэшированный контент, даже если артефакт не найден в вышестоящем реестре или вышестоящий реестр недоступен. Это может привести к отдаче устаревшего контента, если вышестоящий реестр недоступен.",
         "PROXY_REFERRER_API_TIP": "Enable this to proxy OCI 1.1 referrer API requests to the upstream registry.",
-        "PROXY_REFERRER_API_LABEL": "Enable proxy for referrer API"
+        "PROXY_REFERRER_API_LABEL": "Enable proxy for referrer API",
+        "REPOSITORY_FILTER": "Фильтр репозитория",
```
Comment (chlins) [inline src/portal/src/i18n/lang/ru-ru-lang.json]: The seven new i18n keys are added twice in this file: once here under
`SIDE_NAV.PROJECT` (L425-431) and once under the top-level `PROJECT`
section (L1500-1506).

The components resolve them via `PROJECT.REPOSITORY_FILTER_*`, so only
the second block is actually used — this one under `SIDE_NAV.PROJECT`
is dead and doesn't exist in any of the other nine language files
(they all got exactly 7 lines, ru-ru got 14).

Looks like a copy-paste landed in the wrong section. Could you drop
the `SIDE_NAV.PROJECT` block so ru-ru stays consistent with the other
locales?

File: src/portal/src/app/base/left-side-nav/projects/create-project/create-project.component.ts [hunk 2]
```diff
@@ -350,6 +355,60 @@ export class CreateProjectComponent
         }
     }
 
+    validateRepositoryFilter(): void {
```
Comment (chlins) [inline src/portal/src/app/base/left-side-nav/projects/create-project/create-project.component.ts:359]: `validateRepositoryFilter()` and `checkBalancedChars()` here (L358-410)
are verbatim copies of the same methods in
`project-policy-config.component.ts` (L253-305) — ~60 duplicated lines
of validation logic.

Since this validation encodes the contract with the backend matcher
(#23527: kind enum + pattern rules), keeping two copies means they can
silently drift the next time the backend rules change. Could we extract
them into a shared helper (e.g. a small validator util under
`shared/units.ts` or a dedicated `repository-filter.util.ts`) and import
it from both components?

That would also give us a single place to unit-test the edge cases
(invalid regex, unbalanced brackets, escaped `\[` literals).

**Comment by wy65701436** [comment]
cc @AllForNothing could you please help to review?thanks.

File: src/portal/src/i18n/lang/ru-ru-lang.json (hunk 1 above)
Comment (stonezdj) [reply src/portal/src/i18n/lang/ru-ru-lang.json]: There are lots redundant key in the ru-ru-lang.json, create an other issue to track it: https://github.com/goharbor/harbor/issues/23668

File: src/portal/src/app/base/left-side-nav/projects/create-project/create-project.component.ts (hunk 2 above)
Comment (stonezdj) [reply src/portal/src/app/base/left-side-nav/projects/create-project/create-project.component.ts:359]: Updated

**Review by chlins — APPROVED** [review]
lgtm

**Review by wy65701436 — APPROVED** [review]
lgtm
